### PR TITLE
Add FunctionManager

### PR DESCRIPTION
**Pull request by rongrong**
Add FunctionManager and FunctionNamespace, and use FunctionManager instead of FunctionRegistry through out the codebase.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.sql.analyzer.TypeSignatureProvider;
+import com.facebook.presto.sql.tree.QualifiedName;
+
+import java.util.List;
+
+public interface FunctionNamespace
+{
+    void addFunctions(List<? extends SqlFunction> functions);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java]: We name the method below `list` to list the functions in a function namespace, but we name `addFunctions` instead of just `add` to add functions to a namespace. It would be better if we are more consistent here (e.g., `listFunctions` and `addFunctions`).

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    List<SqlFunction> list();
+
+    Signature resolveFunction(QualifiedName name, List<TypeSignatureProvider> parameterTypes);
+
+    WindowFunctionSupplier getWindowFunctionImplementation(Signature signature);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java]: nit: `getAggregateFunctionImplementation`/`getScalarFunctionImplementation` return an implementation, but `getWindowFunctionImplementation` returns a supplier for the implementation. What do you think about naming this method `getWindowFunctionSupplier` or `getWindowFunctionImplementationSupplier`?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    InternalAggregationFunction getAggregateFunctionImplementation(Signature signature);
+
+    ScalarFunctionImplementation getScalarFunctionImplementation(Signature signature);
+
+    boolean canResolveOperator(OperatorType operatorType, Type returnType, List<? extends Type> argumentTypes);
+
+    Signature resolveOperator(OperatorType operatorType, List<? extends Type> argumentTypes)
+            throws OperatorNotFoundException;
+
+    Signature getCoercion(Type fromType, Type toType);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java]: Do we really need both `getCoercion` methods in this interface? (We can get the type signature of a type with `type.getTypeSignature()`).

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java [hunk 4]
```diff
@@ -62,6 +67,19 @@ public Signature(
         this.returnType = requireNonNull(returnType, "returnType is null");
         this.argumentTypes = ImmutableList.copyOf(requireNonNull(argumentTypes, "argumentTypes is null"));
         this.variableArity = variableArity;
+        this.catalog = requireNonNull(catalog, "catalog is null");
+        this.schema = requireNonNull(schema, "schema is null");
+    }
+
+    public Signature(String name,
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]:
```
public Signature(
        String name,
```

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java [hunk 5]
```diff
@@ -109,6 +127,19 @@ public static Signature internalScalarFunction(String name, TypeSignature return
         return new Signature(name, SCALAR, ImmutableList.of(), ImmutableList.of(), returnType, argumentTypes, false);
     }
 
+    public static Signature qualifySignature(Signature signature, String catalog, String schema)
+    {
+        return new Signature(signature.name,
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]:
```
return new Signature(
        signature.name,
```

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+    public void addFunctions(List<? extends SqlFunction> functions)
+    {
+        addFunctions(TEMP_DEFAULT_CATALOG, functions);
+    }
+
+    public synchronized void addFunctions(String catalog, List<? extends SqlFunction> functions)
+    {
+        Map<Boolean, List<SqlFunction>> operatorAndFunctionMap = functions.stream()
+                .collect(Collectors.partitioningBy(function -> isOperator(function.getSignature())));
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: static import `partitioningBy`

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                builder.addAll(namespace.list());
+            }
+        }
+        return builder.build();
+    }
+
+    public Signature resolveFunction(Session session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
+    {
+        try {
+            return operatorNamespace.resolveFunction(name, parameterTypes);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: I think we should have two methods: `resolveFunction` and `resolveOperator` instead of trying to do both in the same method.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        }
+        catch (Exception e) {
+            //do nothing and try to resolve the function using the path
+        }
+        for (SqlPathElement element : session.getPath().getParsedPath()) {
+            FunctionNamespace namespace = functionNamespaces.get(getCatalog(element, session));
+            if (namespace != null) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: If we check against null and continue, that would help us get rid of the nestedness below:

```
if (namespace == null) {
    continue;
}
try {
    Signature signature = namespace.resolveFunction(name, parameterTypes);
    return qualifySignature(signature, getCatalog(element, session), TEMP_DEFAULT_SCHEMA);
}
catch (PrestoException e) {
    if (!e.getErrorCode().equals(FUNCTION_NOT_FOUND.toErrorCode())) {
        throw e;
    }
}
```

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+                }
+            }
+        }
+        throw new PrestoException(FUNCTION_NOT_FOUND, format("Function %s not found in path: %s", name, session.getPath()));
+    }
+
+    public WindowFunctionSupplier getWindowFunctionImplementation(Signature signature)
+    {
+        try {
+            return operatorNamespace.getWindowFunctionImplementation(signature);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: Looks like we do in many of these methods:
- try resolving in the operator namespace
- on exception, try the function namespace

Is it easy to get rid of this behavior? (I don't really remember/know why we check the operator namespace for functions).

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    {
+        try {
+            return operatorNamespace.resolveOperator(operatorType, argumentTypes);
+        }
+        catch (PrestoException e) {
+            if (!(e instanceof OperatorNotFoundException)) {
+                throw e;
+            }
+        }
+        throw new OperatorNotFoundException(
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: Why don't we throw `e` above when it is an instance of `OperatorNotFoundException`? Doesn't it already include the info we need?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    {
+        try {
+            return operatorNamespace.getCoercion(fromType, toType);
+        }
+        catch (PrestoException e) {
+            if (!(e instanceof OperatorNotFoundException)) {
+                throw e;
+            }
+        }
+        throw new OperatorNotFoundException(OperatorType.CAST, ImmutableList.of(fromType), toType);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java (hunk 3 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java]: It's hard to tell at this point since the class is not used. I'll remove the `TypeSignature` one for now. It won't hurt.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java (hunk 8 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: I don't really like how the code looks with handling unresolved functions through exceptions. It would look nicer if the `namespace.resolveFunction` would just return `null` and only FunctionManager would throw.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java (hunk 9 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: I don't really like how the code looks with handling unresolved functions through exceptions. It would look nicer if the `namespace.resolveFunction` would just return `null` and only FunctionManager would throw.

**Comment by rongrong** [comment]
A few changes:
* removed the `operatorNamespace` from `FunctionManager` for now cause it's not obvious from these diffs why this is needed. We can introduce it when necessary.
* Separate the global function namespace out as a special one. Currently this would have no catalog and schema attached to it (the catalog, schema from the resolved function would be null).
* The map of dynamic function namespaces in `FunctionManager` will be a `Map<SqlPathElement, FunctionNamespace>` so each schema can define their own functions.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.featuresConfig = requireNonNull(featuresConfig, "featuresConfig is null");
+        this.globalFunctionNamespace = new StaticFunctionNamespace(typeManager, blockEncodingSerde, featuresConfig);
+        this.dynamicFunctionNamespaces = ImmutableMap.of();
+    }
+
+    public synchronized void addFunctions(String catalog, String schema, List<? extends SqlFunction> functions)
+    {
+        if (!dynamicFunctionNamespaces.containsKey(catalog)) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: Now that `dynamicFunctionNamespaces` is a `Map<SqlPathElement, FunctionNamespace>`, you should fix the various `containsKey` and `get` calls here and below as they pass `String` to those methods instead of `SqlPathElement`.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+import static com.facebook.presto.spi.StandardErrorCode.FUNCTION_NOT_FOUND;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_FOUND;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class FunctionManager
+{
+    private final FunctionNamespace globalFunctionNamespace;
+    private volatile Map<SqlPathElement, FunctionNamespace> dynamicFunctionNamespaces;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: Currently we only have a single subclass (`StaticFunctionNamespace`) of `FunctionNamespace`, and we use instances of  `StaticFunctionNamespace` for the `dynamicFunctionNamespaces` field, which is confusing. Looking at the rest of the patch, do you think we will need a `DynamicFunctionNamespace` class? If we won't, maybe we should make `FunctionNamespace` a class instead of an interface, and use it for both static/dynamic namespaces. Thoughts?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        }
+        return builder.build();
+    }
+
+    public Signature resolveFunction(Session session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
+    {
+        try {
+            return globalFunctionNamespace.resolveFunction(name, parameterTypes);
+        }
+        catch (Exception e) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: As you pointed out offline, I also don't like the way we do resolution through catching exceptions. We should get rid of this.

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java [hunk 15]
```diff
@@ -40,6 +41,8 @@
     private final TypeSignature returnType;
     private final List<TypeSignature> argumentTypes;
     private final boolean variableArity;
+    private final Optional<String> catalog;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]: I don't think it's valid to have a schema without a catalog. This should probably be `Optional<CatalogSchemaName>`

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java (hunk 13 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: Within this three diffs dynamicFunctionNamespaces are not really used. If we really don't want to introduce potentially unnecessary logic, I can change FunctionNamespace to be a class for now (and we can change that later when needed). We can remove the dynamicFunctionNamespaces private variable all together for now.

File: presto-main/src/main/java/com/facebook/presto/metadata/PolymorphicScalarFunctionBuilder.java [hunk 16]
```diff
@@ -143,7 +143,7 @@ public SqlScalarFunction build()
     private static boolean isOperator(Signature signature)
     {
         for (OperatorType operator : OperatorType.values()) {
-            if (signature.getName().equals(FunctionRegistry.mangleOperatorName(operator))) {
+            if (signature.getName().equals(FunctionUtils.mangleOperatorName(operator))) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/PolymorphicScalarFunctionBuilder.java]: static import

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java [hunk 17]
```diff
@@ -202,12 +202,12 @@ public Expression toExpression(Object object, Type type)
             // able to encode it in the plan that gets sent to workers.
             // We do this by transforming the in-memory varbinary into a call to from_base64(<base64-encoded value>)
             FunctionCall fromBase64 = new FunctionCall(QualifiedName.of("from_base64"), ImmutableList.of(new StringLiteral(VarbinaryFunctions.toBase64((Slice) object).toStringUtf8())));
-            Signature signature = FunctionRegistry.getMagicLiteralFunctionSignature(type);
+            Signature signature = FunctionUtils.getMagicLiteralFunctionSignature(type);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java]: static import (also the ones at L209 and L210)

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArrayHashCodeOperator.java [hunk 18]
```diff
@@ -98,7 +98,7 @@
     public static class BenchmarkData
     {
         @Param({"$operator$hash_code", "old_hash", "another_hash"})
-        private String name = FunctionRegistry.mangleOperatorName(HASH_CODE);
+        private String name = FunctionUtils.mangleOperatorName(HASH_CODE);
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArrayHashCodeOperator.java]: static import

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java [hunk 19]
```diff
@@ -49,7 +52,8 @@ public Signature(
             @JsonProperty("longVariableConstraints") List<LongVariableConstraint> longVariableConstraints,
             @JsonProperty("returnType") TypeSignature returnType,
             @JsonProperty("argumentTypes") List<TypeSignature> argumentTypes,
-            @JsonProperty("variableArity") boolean variableArity)
+            @JsonProperty("variableArity") boolean variableArity,
+            @JsonProperty("catalog") Optional<CatalogSchemaName> catalogSchemaName)
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]: `@JsonProperty("catalog")` -> ` @JsonProperty("catalogSchemaName")` ?

File: presto-main/src/test/java/com/facebook/presto/metadata/AbstractMockMetadata.java [hunk 20]
```diff
@@ -69,7 +69,6 @@ public boolean isAggregationFunction(QualifiedName name)
         throw new UnsupportedOperationException();
     }
 
-    @Override
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/metadata/AbstractMockMetadata.java]: why is this removed?

File: presto-main/src/test/java/com/facebook/presto/metadata/TestFunctionRegistry.java [hunk 21]
```diff
@@ -293,6 +293,14 @@ public void testResolveFunctionForUnknown()
                 .failsWithMessage("Could not choose a best candidate operator. Explicit type casts must be added.");
     }
 
+    private FunctionRegistry createFunctionRegistry(TypeRegistry typeManager)
+    {
+        BlockEncodingManager blockEncodingManager = new BlockEncodingManager(typeManager);
+        FeaturesConfig featuresConfig = new FeaturesConfig();
+        FunctionManager functionManager = new FunctionManager(typeManager, blockEncodingManager, featuresConfig);
+        return new FunctionRegistry(typeManager, blockEncodingManager, new FeaturesConfig(), functionManager);
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/metadata/TestFunctionRegistry.java]: We create multiple `FeaturesConfig`s in this method (another one is created on L299).

**Comment by dain** [comment]
I would like to review this, so please don't merge yet

**Comment by dain** [comment]
The author of these commits should be `Cole Bowden <bowdencm@fb.com>`

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+        }
+        throw new PrestoException(FUNCTION_NOT_FOUND, format("Function %s not found in path: %s", name, session.getPath()));
+    }
+
+    public WindowFunctionSupplier getWindowFunctionSupplier(Signature signature)
+    {
+        try {
+            return globalFunctionNamespace.getWindowFunctionSupplier(signature);
+        }
+        catch (PrestoException e) {
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: For context, naming it `ignored` (or `expected) suppresses the IntelliJ warning about unused exception field.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.Set;
+
+import static com.facebook.presto.metadata.FunctionKind.SCALAR;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public final class FunctionUtils
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java]: I feel like this should have a more specific name like: `FunctionSignatureUtils` or `SignatureMangler`.... not sureI like those two, but something more specific than "utilities for functions".

**Comment by rongrong** [comment]
@dain The author is no longer Cole because these are pretty much all reimplemented. The structure and content of all commits you are seeing now are dramatically changed compare to the original PR. There were a lot of git reset due to restructure of diffs which caused the name change. Not sure what's your motivation for put Cole's name behind it. I can change them to Cole, but these commits are probably 80%+ done by me at the moment and I should be responsible for them.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java (hunk 23 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java]: This is one of the commits that's more or less still kept its look from the original PR. I think the intention is to move the static methods out so other classes don't reference `FunctionRegistry` directly. However, if you look at the methods here, only `mangleOperatorName` is widely used. All others can potentially be moved somewhere else. I was debating whether I want to do that now, but figured that maybe it's better to do it later once `FunctionRegistry` is removed and the logic for operators and functions are sorted out into proper places.

**Comment by rongrong** [comment]
For the change to `Signature`, the only reason it's here is that adding Catalog and Schema was the first commit in the original PR. I agree that it's not related. We can move that out.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+{
+    // hack: java classes for types that can be used with magic literals
+    private static final Set<Class<?>> SUPPORTED_LITERAL_TYPES = ImmutableSet.of(long.class, double.class, Slice.class, boolean.class);
+    static final String MAGIC_LITERAL_FUNCTION_PREFIX = "$literal$";
+    private static final String OPERATOR_PREFIX = "$operator$";
+
+    private FunctionUtils() {}
+
+    public static Signature getMagicLiteralFunctionSignature(Type type)
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java]: Ok did a bunch of digging...  I suggest we move the *literal* related functions to `LiteralEncoder`.  I also suggest we hide the "magic" literal stuff a bit better, by introducing this method:
```
    private static Optional<Signature> resolveLiteralFunction(QualifiedName name, List<TypeSignatureProvider> parameterTypes, TypeManager typeManager)
    {
        if (!name.getSuffix().startsWith(MAGIC_LITERAL_FUNCTION_PREFIX)) {
            return Optional.empty();
        }

        // extract type from function name
        String typeName = name.getSuffix().substring(MAGIC_LITERAL_FUNCTION_PREFIX.length());

        // lookup the type
        Type type = typeManager.getType(parseTypeSignature(typeName));

        // verify we have one parameter of the proper type
        checkArgument(parameterTypes.size() == 1, "Expected one argument to literal function, but got %s", parameterTypes);

        return Optional.of(getMagicLiteralFunctionSignature(type));
    }
```

Then in `FunctionRegistry.resolveFunction`, we do something like:
```
        return resolveLiteralFunction(name, parameterTypes, typeManager)
                .orElseThrow(() -> new PrestoException(FUNCTION_NOT_FOUND, message));
```

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java (hunk 23 above)
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java]: If we can make the literal changes I suggested, I think we can rename this to something like `OperatorSignatureUtils`

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        return mangleOperatorName(operatorType.name());
+    }
+
+    public static String mangleOperatorName(String operatorName)
+    {
+        return OPERATOR_PREFIX + operatorName;
+    }
+
+    public static Type typeForMagicLiteral(Type type)
```
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java]: It's used in `LiteralEncoder` as well.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java (hunk 24 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java]: If this is just a private function in `FunctionRegistry` not sure how much "hiding" it adds. A practical problem with this is that there's an error about `message` being not `final`. Do you think this is really necessary?

**Comment by rongrong** [comment]
@dain addressed your comments other than the `resolveLiteralFunction` suggestion. Let me know what you think.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+class FunctionNamespace
+{
+    private final FunctionRegistry functionRegistry;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionNamespace.java]: Maybe shorten this to `registry` since it's the only field

File: presto-main/src/main/java/com/facebook/presto/metadata/OperatorSignatureUtils.java [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.presto.spi.function.OperatorType;
+import com.google.common.annotations.VisibleForTesting;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public final class OperatorSignatureUtils
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/OperatorSignatureUtils.java]: Method references will look nicer if we name this `OperatorSignatures`

File: presto-main/src/main/java/com/facebook/presto/metadata/OperatorSignatureUtils.java [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+
+    public static String mangleOperatorName(String operatorName)
+    {
+        return OPERATOR_PREFIX + operatorName;
+    }
+
+    @VisibleForTesting
+    public static OperatorType unmangleOperator(String mangledName)
+    {
+        checkArgument(mangledName.startsWith(OPERATOR_PREFIX), "%s is not a mangled operator name", mangledName);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/OperatorSignatureUtils.java]: It's better to put the variable part last in error messages (easier to search for, etc.):
```java
"not a mangled operator name: %s"

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java [hunk 29]
```diff
@@ -197,18 +205,56 @@ public Expression toExpression(Object object, Type type)
...
+    public static boolean isSupportedLiteralType(Type type)
+    {
+        return SUPPORTED_LITERAL_TYPES.contains(type.getJavaType());
+    }
+
+    public static Signature getMagicLiteralFunctionSignature(Type type)
+    {
+        TypeSignature argumentType = typeForMagicLiteral(type).getTypeSignature();
+
+        return new Signature(MAGIC_LITERAL_FUNCTION_PREFIX + type.getTypeSignature(),
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java]: I'd format this like
```java
String name = MAGIC_LITERAL_FUNCTION_PREFIX + type.getTypeSignature();
return new Signature(name, SCALAR, type.getTypeSignature(), argumentType);

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java [hunk 30]
```diff
@@ -197,18 +205,56 @@ public Expression toExpression(Object object, Type type)
...
+            return BIGINT;
+        }
+        if (clazz == double.class) {
+            return DOUBLE;
+        }
+        if (!clazz.isPrimitive()) {
+            if (type instanceof VarcharType) {
+                return type;
+            }
+            else {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java]: Else is redundant

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java [hunk 31]
```diff
@@ -197,18 +205,56 @@ public Expression toExpression(Object object, Type type)
...
+                return VARBINARY;
+            }
+        }
+        if (clazz == boolean.class) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java]: Move this above so it's with the other primitive class checks

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java [hunk 32]
```diff
@@ -197,18 +205,56 @@ public Expression toExpression(Object object, Type type)
...
+
+        return new Signature(MAGIC_LITERAL_FUNCTION_PREFIX + type.getTypeSignature(),
+                SCALAR,
+                type.getTypeSignature(),
+                argumentType);
+    }
+
+    private static Type typeForMagicLiteral(Type type)
+    {
+        Class<?> clazz = type.getJavaType();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java]: Inline the second assignment (separate lines looks strange because it's short, and it's better to assign variables only once if possible)
```java
Class<?> clazz = Primitives.unwrap(type.getJavaType());

Comment (electrum) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java]: Actually, this seems wrong. We shouldn't ever have boxed types as the Java type for a class. This should probably be
```java
Class<?> clazz = type.getJavaType();
checkArgument(!Primitives.isWrapperType(clazz), "Java class for type %s is wrapper %s", type, clazz.getName());

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 33]
```diff
@@ -0,0 +1,113 @@
...
+
+@ThreadSafe
+public class FunctionManager
+{
+    private final FunctionNamespace globalFunctionNamespace;
+    private final FunctionInvokerProvider functionInvokerProvider;
+
+    public FunctionManager(TypeManager typeManager, BlockEncodingSerde blockEncodingSerde, FeaturesConfig featuresConfig)
+    {
+        FunctionRegistry functionRegistry = new FunctionRegistry(typeManager, blockEncodingSerde, featuresConfig, this);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: This commit does not compile since the `FunctionRegistry` constructor was not changed

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 34]
```diff
@@ -0,0 +1,113 @@
...
+        this.globalFunctionNamespace = new FunctionNamespace(functionRegistry);
+        this.functionInvokerProvider = new FunctionInvokerProvider(functionRegistry);
+        if (typeManager instanceof TypeRegistry) {
+            ((TypeRegistry) typeManager).setFunctionManager(this);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: This also won't compile

File: presto-main/src/main/java/com/facebook/presto/sql/gen/BytecodeGeneratorContext.java [hunk 35]
```diff
@@ -33,15 +33,15 @@
     private final Scope scope;
     private final CallSiteBinder callSiteBinder;
     private final CachedInstanceBinder cachedInstanceBinder;
-    private final FunctionRegistry registry;
+    private final FunctionManager registry;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/BytecodeGeneratorContext.java]: Rename to `manager` or `functionManager` in all places. Search for `Manager registry` to find the instances

**Comment by sopel39** [comment]
Is there a plan to land it soon?

**Comment by rongrong** [comment]
@sopel39 I'll land this after the current freeze for release.

File: presto-main/src/main/java/com/facebook/presto/metadata/OperatorSignatureUtils.java (hunk 27 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/OperatorSignatureUtils.java]: We normally do static import of static methods so the class name would not appear at the site of method reference. I don't have strong opinions of this class name though. I have a strong feeling that this class would go away some day anyways. I can rename if it would make you happier.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java (hunk 29 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java]: Normally when I do refactoring I don't change the original code. I found this easier for code review. (You can safely ignore then content when reviewing as long as it still compiles). But if we also want to make the code nicer, I'm open to it. So far I didn't change the name or implementation of logic refactored. Let me know if you think otherwise.

**Comment by rongrong** [comment]
Updated so each individual commit builds successfully. Also updated the commit messages.
